[2단계 - 장바구니] 루나(오신정) 미션 제출합니다.

ShinjungOh opened the pull request:
# 🛒 장바구니

안녕하세요 해먼드! 장바구니 미션 2단계를 완성해서 리뷰 요청 드립니다.
이번에도 잘 부탁드려요! 🙇‍♀️🍀
시간이 부족해서 `가장 할인 효과가 큰 조합을 자동으로 선택`, `장바구니 선택 상태를 로컬스토리지에 저장`하는 기능을 미처 구현하지 못했어요. 😭
리팩토링 하면서 하나씩 다 반영하려고 합니다!!

## 📦 장바구니 미션

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계

- 클라이언트 상태를 효과적으로 모델링하고 관리할 수 있다.
- Jest, React Testing Library(RTL)를 활용하여 주요 기능에 대한 테스트를 작성할 수 있다.

### 2단계

- 상태를 중복 없이 구성하고, 계산이 가능한 값은 파생 상태로 도출하여 관리할 수 있다.
- 함수(컴포넌트, 훅, util)가 단일한 책임만 갖도록 분리하고, 그 책임에 맞는 명확한 인터페이스를 정의할 수 있다.

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [ ] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [ ] 테스트 케이스를 모두 작성했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
	- 배포 링크 기입: [🔗 배포 사이트](https://shinjungoh.github.io/react-shopping-cart/)

### 리뷰 요청 & 논의하고 싶은 내용

### 테스트 통과


![CleanShot 2025-06-09 at 17 17 55@2x](https://github.com/user-attachments/assets/db9d6fde-6d15-4763-85d7-71bcf3dd79ac)


### 1) 상태 관리 설계 의도

Context를 통해 받아온 원천 데이터(장바구니, 쿠폰 목록)를 이용해, 
하위 컴포넌트에서 필요한 파생 상태(할인 금액, 배송비 등)를 계산하여 사용하도록 했어요.

저에게는 바텀업 방식의 구현이 어려워서 탑다운 방식으로 진행했는데요, 
상태 관리가 복잡해서 우선 OrderConfirmPage에서 기능과 스타일을 다 정의해두고, 
구현이 완료되었을 때 하나씩 분리하는 방식으로 리팩토링을 진행했어요. 
쿠폰 할인값 계산, 배송비 계산, 쿠폰 데이터 패칭 등을 `커스텀 훅`으로 분리하고, 순수함수인 계산 관련 로직은 `유틸 함수`로 분리했어요.
비즈니스 로직은 관심사에 맞게 훅과 유틸 함수로 분리하고, 컴포넌트는 UI에 관한 책임만 가지도록 했어요.


### 2) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

1. `OrderConfirmPage` 컴포넌트에서 배송비 계산, 쿠폰 선택, 할인 계산 등 모든 로직을 직접 관리하고 있어서 컴포넌트가 비대해지고, 쿠폰 모달 등 UI도 분리되어 있지 않아서 복잡한 코드가 되었어요.
이에 따라 어디까지를 커스텀 훅/유틸 함수로로 분리해야 할지, 상태의 소유권을 누가 가져야 하는지에 대해 고민을 하게 되었어요.
훅을 만드는 경우에는 훅의 반환값은 어떤 수준으로 추상화해야 하는가에 대한 생각도 해보게 되었어요. 
로직을 순수 함수로 분리하니 테스트 코드를 작성하기 쉬워져서, 적절한 추상화가 중요하다는 것을 깨달았어요.


2. 1단계 미션에서 피드백 주신 내용을 최대한 반영하려고 했어요.
https://github.com/woowacourse/react-shopping-cart/pull/342#discussion_r2134315927
https://github.com/woowacourse/react-shopping-cart/pull/342#discussion_r2134262347

이 두가지를 아직 챙기지 못했는데, 리팩토링 때 반영하려고 합니다.


### 3) 이번 리뷰를 통해 논의하고 싶은 부분

1. `훅의 반환값은 UI에 필요한 최소한의 정보만 포함하고, 내부 구현은 외부에 드러나지 않도록 캡슐화한다`는 요구사항을 충족하고 있는지 궁금해요.
2. `OrderConfirmPage` 컴포넌트에 여러 훅에서 불러온 값들이 있는데, 여기에서 사용되는 로직도 `useOrderConfirm` 훅 등으로 분리해주는 것이 좋을까요? 테스트 코드를 작성하려면 통합된 훅을 만드는 것이 좋아보이는데, 훅이 너무 많아지는 것 같아서 궁금해졌어요! 

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 상태관리

- 원본상태/파생상태를 적절히 구분하여 선언하였나요?
- React state 를 불필요하게 선언한 부분은 없나요?
- 상태 관리 로직의 책임이 적절히 응집/분리되었나요? (ex. reducer, hook, Context)

### 2. 역할과 책임

- 단일 책임 원칙(SRP)에 맞게 컴포넌트/함수가 분리되었나요?
- 컴포넌트/함수의 인터페이스가 구현에 의존하지 않고, 책임에 맞게 설계되었나요?

### 3. 테스트

- 쿠폰 기능에 대한 다양한 시나리오가 충분히 테스트되었나요?
- 함수의 책임을 분리하고, 순수성을 유지하는 등 테스트하기 좋은 코드를 작성하였나요?

ShinjungOh committed:
docs: README 기능 요구사항 작성

ShinjungOh committed:
feat: 모달 컴포넌트 추가

ShinjungOh committed:
feat: 쿠폰 정보 및 타입 추가

ShinjungOh committed:
refactor: 쿠폰 정보 상수명 대문자 스네이크 케이스로 변경

ShinjungOh committed:
feat: 결제 금액 확인 페이지 UI 구현

ShinjungOh committed:
feat: 쿠폰 선택 모달 UI 구현

ShinjungOh committed:
feat: 모달의 쿠폰 세부 내용 UI 구현

ShinjungOh committed:
feat: 주문 확인 페이지 UI 구현

ShinjungOh committed:
feat: 쿠폰 데이터 및 타입 지정

ShinjungOh committed:
feat: 쿠폰 적용 모달 날짜 및 시간 포맷팅 유틸 추가

ShinjungOh committed:
test: 유틸 함수 테스트 작성
- dateTimeFormatter, cartCalculations 테스트

ShinjungOh committed:
feat: 쿠폰 조회 API 함수 생성 및 쿠폰 데이터 패칭 훅 구현

ShinjungOh committed:
feat: 제주도 및 도서 산간 지역 체크 시 배송비 추가 로직 구현

ShinjungOh committed:
feat: 장바구니 선택 상품을 주문 확인 페이지에 표시

- CartPage에서 선택한 상품들의 전체 정보를 location.state를 이용해 OrderConfirmPage로 전달
- useCart 훅에 selectedProducts 추가

ShinjungOh committed:
feat: 쿠폰 선택 토글 및 유효성 검증 기능 추가

- 쿠폰 만료일/시간대별 사용 가능 여부 체크
- 선택된 쿠폰 할인 금액 계산 로직 추가
- 쿠폰 최대 2개 선택 가능하도록 제한

ShinjungOh committed:
feat: 쿠폰 타입별 할인 계산 로직 구현

ShinjungOh committed:
refacotr: couponCalculations 파일로 쿠폰 할인 계산 로직 분리

ShinjungOh committed:
feat: 무료 배송 쿠폰 적용 시 도서 산간 지역 배송비 무료 처리

ShinjungOh committed:
refactor: CartOrderItem 컴포넌트 분리

ShinjungOh committed:
refactor: DeliveryOptions 컴포넌트 분리

ShinjungOh committed:
refactor: CouponModal 컴포넌트 분리

ShinjungOh committed:
refactor: 배송비 관련 로직 useShippingFee 훅으로 분리

ShinjungOh committed:
refactor: couponAvailability 유틸 함수로 쿠폰 유효성 검증 로직 분리

ShinjungOh committed:
refactor: 쿠폰 선택 로직을 useCouponSelection 훅으로 분리

ShinjungOh committed:
test: 쿠폰 적용 유효성 검사 테스트 추가

ShinjungOh committed:
test: 쿠폰 할인 적용 테스트 추가

ShinjungOh committed:
test: CartPage 테스트가 통과하도록 코드 수정

ShinjungOh committed:
refactor: 불필요 타입 제거

ShinjungOh committed:
refactor: as 타입 단언으로 타입 에러 해결 및 불필요 함수를 값으로 변경

ShinjungOh committed:
refactor: SelectBox의 props를 cartItem에서 id로 변경

ShinjungOh committed:
test: waitFor 제거 및 DataProvider 모킹으로 API 호출 에러 해결

JunilHwang commented on src/apis/coupon.ts:
```diff
@@ -0,0 +1,10 @@
+import { apiClient } from './apiClient';
+
+export const getCoupons = async () => {
+  try {
+    return await apiClient<Coupon[]>('/coupons');
+  } catch (error) {
+    throw new Error('쿠폰 목록을 불러오는 중 에러 발생');
+  }
```
try/catch가 없는게 더 좋지 않을까요!?
지금은 에러가 다 손실되는 것 같아요..!

JunilHwang commented on src/apis/cart.ts:
```diff
@@ -1,7 +1,7 @@
-export const getCartItems = async (): Promise<CartItemsResponse> => {
+export const getCartItems = async () => {
   try {
     return await apiClient<CartItemsResponse>('/cart-items?page=0&size=20');
```
```tsx
return await apiClient<CartItemsResponse>('/cart-items', {
  query: {
    page: 0,
    size: 20,
  }
});
```
이렇게 사용할 수 있도록 추상화해주면 어떨까요!?

ShinjungOh replied:
리뷰주신대로 추상화 했습니다!

[refactor: apiClient 함수 추상화](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/86ca474492ce06f0f98d0ba8e9a5e49f3bd86b9d)

JunilHwang commented on test/utils/cartCalculations.test.ts:
```diff
@@ -0,0 +1,131 @@
+import { describe, it, expect } from 'vitest';
+import {
+  filterCheckedItems,
+  calculateTotalPrice,
+  calculateTotalQuantity,
+  calculateShippingFee,
+  getCartDescription,
+} from '../../src/utils/cartCalculations';
+import { CartProduct } from '../../src/types/cart';
+import {
+  SHIPPING_FEE,
+  FREE_SHIPPING_FEE,
+  SHIPPING_FEE_THRESHOLD,
+} from '../../src/constants/cartConfig';
```
지금은 테스트와 테스트 대상이 코드 사이의 거리가 무척 멀어요.
루나의 PR에서 직접 리뷰를 해보시면, 현재 테스트에서 실행하는 함수의 내용이 궁금해서 탐색해보려고 하면 꽤 번거로운 것을 알 수 있답니다 ㅎㅎ
즉, 코드에 대한 응집도가 낮다고 생각해요.

이 문제를 어떻게 해결할 수 있을까요!?

JunilHwang commented on src/components/Cart/CartFooter.tsx:
```diff
@@ -1,19 +1,35 @@
+    if (value > 0) {
+      return `-${couponDiscount}`;
+    }
+  };
```
`isZero` 라는 이름만 보면 함수의 결과가 boolean을 반환한다고 생각했는데, discount를 반환하네요!?
이렇게 쓰이면 어떨까요?

```suggestion
  const adjustedDiscount = value !== 0 ? -couponDiscount : 0;
```

그냥 값으로 표현하는거죠..!

ShinjungOh replied:
함수명을 반환값의 타입과 부적절하게 지었네요..🥲 
분기 조건인 '0인지 아닌지'를 생각하다가 함수명에도 `is-`를 붙여버린 것 같아요.
피드백 주신 것처럼 로직을 값으로 변경했어요!

[refactor: 쿠폰 할인 금액의 음수 처리 로직 단순화](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/2326f80ee4243f4a999278164fb9284ba5ba45dc)

JunilHwang commented on src/components/Cart/CartInfo.tsx:
```diff
@@ -0,0 +1,31 @@
+
+function CartInfo({ description, style }: CartInfoProps) {
+  return (
+    <Container style={style}>
```
이모션을 사용하고 있으니까, style대신에 css로 표현하면 좋을 것 같아요.

```suggestion
    <Container css={style}>
```

style대신에 css를 사용하는게 어떤 차이가 있는걸까요!?

ShinjungOh replied:
찾아보니 주요 차이점은 css를 사용할 경우에 동적 스타일링이 가능하고 의사 클래스, 미디어 쿼리 등의 요소를 사용할 수 있다는 점이 있네요! 
`css`를 사용하는 것으로 개선했어요!

다음은 AI에게 물어본 내용입니다.

### style prop vs css prop

1. style prop (인라인 스타일)

`<div style={{ marginTop: '32px' }}>`
  - React의 기본 인라인 스타일
  - JavaScript 객체로 스타일 정의
  - camelCase 속성명 사용 (marginTop)
  - 단위 자동 추가 (숫자만 쓰면 px 자동 추가)
  - 의사 클래스/요소 사용 불가 (:hover, ::before 등)
  - 미디어 쿼리 사용 불가

2. css prop (Emotion CSS)

`<div css={css`margin-top: 32px;`}>`
  - Emotion의 강력한 스타일링 기능
  - CSS 문법 그대로 사용
  - kebab-case 속성명 사용 (margin-top)
  - 의사 클래스/요소 사용 가능
  - 미디어 쿼리 사용 가능
  - 동적 스타일링 가능
  - 더 나은 성능 (스타일 캐싱)


[refactor: CartInfo에 Emotion css prop 적용으로 스타일링 개선](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/853dc09c1375cd4c69e020e664334db2b41e6d0c)

JunilHwang commented on src/components/Cart/CartOrderItem.tsx:
```diff
@@ -0,0 +1,70 @@
+            <ProductQuantity>{product.quantity}개</ProductQuantity>
+          </ProductInfo>
+        </ProductSection>
+      ))}
```
흠.. CartProduct가 product로 표현되고 있는데, 그러다보니 `product.product` 처럼 중복 표현이 쓰이고 있어요. 차라리 cart 라고 표현해주면 어떨까요?

ShinjungOh replied:
`product`로 받아오던 props 이름을 `cart`로 변경했어요.

[refactor: CartOrderItem의 props명을 product에서 cart로 변경](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/7069f975b573953343fc7662039a09c3d4308160)

JunilHwang commented on src/components/Cart/CartPrice.tsx:
```diff
@@ -1,13 +1,22 @@
+
+  if (typeof price === 'undefined') {
+    price = 0;
+  }
```
```suggestion
function CartPrice({ title, price: originPrice = 0, variant = 'default' }: CartPriceProps) {
  const price = Number(String(originPrice).replaceAll(',', ''));
```
이렇게 표현해주면 어떨까요?

ShinjungOh replied:
제안해주신 것처럼 변경 완료했어요!

[refactor: CartPrice의 price prop 타입을 유연하게 변경](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/91af7d51ca232faef3c02c52730ba2fff22dcb59)

JunilHwang commented on src/components/Modal/hooks/useFocusTrap.ts:
```diff
@@ -0,0 +1,77 @@
+    return () => {
+      document.body.style.overflow = originalStyle;
+    };
+  }, [initialFocusRef]);
```
initialFocusRef가 변하는게 맞을까요..!?

JunilHwang commented on src/components/Modal/hooks/useFocusTrap.ts:
```diff
@@ -0,0 +1,77 @@
+  onEscape?: () => void;
+};
+
+export function useFocusTrap<T extends HTMLElement = HTMLElement>({
```
이 훅에 대한 테스트는 없어도 무방한걸까요!?

JunilHwang commented on src/hooks/useCart.ts:
```diff
@@ -18,6 +18,9 @@ export const useCart = () => {
   });
 
+  const selectedProducts =
+    cartItems?.content?.filter((item) => checkedItems.includes(item.id)) || [];
```
useCartData 가 useData를 호출하고 있고, useData 내부에서는 useEffect를 실행하고 있는데요,
이게 단일 책임 원칙에 벗어나는 모습이라고 생각해요.
data만 가져와서 반환하는 훅이 하나 더 필요하지 않을까요?
데이터를 업데이트하는 것과 가져오는 것이 분리되어야 할 것 같아요!

JunilHwang replied:
그리고 이 훅 내부에서 호출하는 다른 훅에서 결국 cartItems를 필요로 해요.
그냥 useCartData를 각각의 훅에서 호출해서 사용하면 되지 않을까요!?

종합해보자면 지금은 context를 사용하고 있으나 활용을 제대로 못하고 있는 모습이라고 생각해요..!

ShinjungOh replied:
context를 활용해 props drilling을 해결하고, 각 훅에서 직접 context로 데이터를 조회하도록 개선했어요!
`CartList`가 직접 필요한 데이터와 로직을 가져와서 사용하고,
`CartMain`도 직접 필요한 훅을 사용하도록 했습니다.

[refactor: Context 활용해 props drilling 해결 및 데이터 흐름 개선](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/37a2212822ff662ba0939e3500ef84066c16149f)

JunilHwang commented on src/hooks/useCartUI.ts:
```diff
@@ -6,7 +6,7 @@ interface UseCartUIProps {
+  const descriptionMessage = getCartDescription(itemCount);
   const isDisabled = checkedItemsCount === 0;
 
   return {
```
여기서 정의된 상태가 서로 연관이 있는가? 라고 하면 그게 아닌 것 같아요.
가령, itemCount는 descriptionMessage에서만 필요하고
checkedItemsCount는 disabled 를 만들 때 필요합니다.

차라리 여기서 쓰이는 로직을 각각의 컴포넌트에서 이를 정의해서 사용해주면 어떨까요!?

ShinjungOh replied:
관련성이 없는 useCartUI 내부의 로직을 useCart로 이동시켰어요!

[refactor: 관련성이 없는 useCartUI 내부의 로직을 useCart로 이동](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/296294f4d720654419919a460e7574cb552f8d45)

JunilHwang commented on src/hooks/useCouponData.ts:
```diff
@@ -0,0 +1,12 @@
+import { getCoupons } from '../apis/coupon';
+
+export const useCouponData = () => {
+  const { data: coupons, refetch } = useData<Coupon[]>({
+    fetcher: getCoupons,
+    name: 'coupons',
+  });
+
+  return { coupons, refetch };
```
지금은 컨텍스트에 하나의 데이터만 담을 수 있답니다!
한 페이지에 여러 종류의 데이터가 필요하면 어떻게 해야 좋을까요?

JunilHwang commented on src/hooks/useCouponSelection.ts:
```diff
@@ -0,0 +1,72 @@
+  applyCoupons: () => void;
+}
+
+export const useCouponSelection = (): UseCouponSelectionReturn => {
```
리턴타입은 정의할 필요가 없어보여요..!
어차피 자연스럽게 추론된답니다.

ShinjungOh replied:
불필요한 리턴 타입을 제거했어요~!

[refactor: 불필요 리턴 타입 제거](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/69128b0b180cbd3463e07829f390e7ca28598c16)

JunilHwang commented on src/hooks/useCouponSelection.ts:
```diff
@@ -0,0 +1,72 @@
+export const useCouponSelection = (): UseCouponSelectionReturn => {
+  const [selectedCoupons, setSelectedCoupons] = useState<Coupon[]>([]);
+  const [tempSelectedCoupons, setTempSelectedCoupons] = useState<Coupon[]>([]);
+  const [isModalOpen, setIsModalOpen] = useState(false);
```
setIsModalOpen이 상태로 관리될 필요가 있을까요? 제 생각에는 아예 없어도 될 것 같아요..!
```tsx
const modalEnabled = tempSelectedCoupons.length > 0
```

ShinjungOh replied:
`isModalOpen`이라는 상태를 제거하고, `tempSelectedCoupons`의 null 여부로 모달의 상태를 판단하도록 개선했어요.
모달의 열림/닫힘 상태와 임시 쿠폰 선택을 하나의 상태로 관리하도록 했어요.

```
tempSelectedCoupons === null  //모달 닫힘
tempSelectedCoupons !== null //모달 열림
```

[refactor: 쿠폰 모달 여닫는 불필요 상태 로직을 값으로 개선](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/5c61b99f325585fb10307517376919705ddb3648)

ShinjungOh replied:
관련 코드를 점진적으로 개선해나가고 있는데, CouponModal이 자체 상태를 가지도록 개선하니
`useCouponSelection` 훅이 불필요해져서 제거했어요. 
개선된 과정에서의 커밋도 같이 남깁니다!

[refactor: 쿠폰 모달 상태 관리 구조 개선](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/f654e76472555af34d82d0af6c6ed9e4ba6babde)

JunilHwang commented on src/hooks/useCouponSelection.ts:
```diff
@@ -0,0 +1,72 @@
+    toggleCouponSelection,
+    isCouponSelected,
+    applyCoupons,
+  };
```
모달 내부에서 관리해야할 데이터와 바깥에서 관리야할 데이터가 혼재되어 있어요.
CountModal이 자체 상태를 가지고 있고,
Modal에서 Confirm이 이뤄지는 순간 onConfirm 혹은 onSuccess 같은걸 토대로 바깥에 알리고
바깥에서는 이를 기반으로 실제 쿠폰에 반영하도록 하면 되지 않을까요!?

ShinjungOh replied:
[아래 코멘트](https://github.com/woowacourse/react-shopping-cart/pull/405#discussion_r2143809980)에 답변 남겼습니다! 
 
[refactor: 쿠폰 모달 상태 관리 구조 개선](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/f654e76472555af34d82d0af6c6ed9e4ba6babde)

JunilHwang commented on src/hooks/useShippingFee.ts:
```diff
@@ -0,0 +1,49 @@
+}: UseShippingFeeProps): UseShippingFeeReturn => {
+  const [remoteArea, setRemoteArea] = useState(false);
+
+  const baseShippingFee = subtotal >= SHIPPING_FEE_THRESHOLD ? 0 : 3000;
```
3000도 상수로 관리되면 좋을 것 같아요!

그리고 배송비가 3000원이 아니거나 혹은 무료 배송비에 대한 조건을 기획자가 설정할 수 있다고 가정해본다면 어떤 방식으로 인터페이스가 구성이 되어야 좋을까요!?

ShinjungOh replied:
상수로 분리를 해두었는데 적용이 되어있지 않았네요. 상수 적용시켰어요!

[refactor: 배송비 관련 상수를 코드에 적용](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/fddcda3b930967eabd9ba20bd699ccb42595afae)

JunilHwang commented on src/components/Modal/CouponModal.tsx:
```diff
@@ -0,0 +1,161 @@
+  tempCouponDiscount,
+}: CouponModalProps) {
+  return (
+    <Modal position="center" width="90%" title="쿠폰을 선택해 주세요" onClose={onClose}>
```
앞선 리뷰를 참고해주세요.
이 모달이 자체 상태를 가지고 있어야 좋을 것 같아요.
무엇보다, 지금은 임시 쿠폰 정보가 변경될 때 불필요하게 전체적으로 렌더링이 발생하는 현상을 보이고 있어요.
모달 내부에서 보여지는 UI에 변경사항이 있으면 모달만 렌더링 하도록 만들기 위해서 어떻게 해야 좋을까요?

ShinjungOh replied:
CouponModal이 자체적으로 임시 선택 상태를 관리하도록 개선했어요.
그 과정에서 `useCouponSelection` 훅이 불필요해져서 제거했어요.
모달 열림/닫힘 상태를 `OrderConfirmPage`에서 직접 관리하고, 모달 내부 상태 변경 시에 불필요한 리렌더링을 방지하도록 했어요!

[refactor: 쿠폰 모달 상태 관리 구조 개선](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/f654e76472555af34d82d0af6c6ed9e4ba6babde)

JunilHwang commented on src/components/Modal/Modal.tsx:
```diff
@@ -0,0 +1,71 @@
+  onClose: () => void;
+};
+
+function Modal({ width, height, position = 'center', title, onClose, children, size }: ModalProps) {
+  const customWidth = position === 'center' ? width : '100%';
+  const closeButtonRef = useRef<HTMLButtonElement>(null);
+
+  const { containerRef, handleKeyDown } = useFocusTrap({
+    initialFocusRef: closeButtonRef as React.RefObject<HTMLButtonElement>,
+    onEscape: onClose,
+  });
+
+  const handleClickOverlay = () => {
+    onClose();
+  };
+
+  return (
+    <Layout onClick={handleClickOverlay}>
```
onClose는 close가 발생했을 때 이를 밖에 알리는 이벤트입니다.
그런데 지금은 모달이 close 기능을 수행하고 있지 않아요.
onClose 라는 이름이 적절한걸까요?

JunilHwang commented on src/constants/cartConfig.ts:
```diff
@@ -1,7 +1,9 @@
 export const SHIPPING_FEE = 3_000;
 
-export const SHIPPING_FEE_THRESHOLD = 100_000;
+export const REMOTE_SHIPPING_FEE = 3_000;
```
이미 3000이 상수로 정의되어 있었군요..!

ShinjungOh replied:
상수를 만들어놓고 제대로 활용하지 못한 곳이 있었네요..!
상수 처리 했습니다.

[refactor: 배송비 관련 상수를 코드에 적용](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/fddcda3b930967eabd9ba20bd699ccb42595afae)

JunilHwang commented on src/pages/CartPage.tsx:
```diff
@@ -20,6 +20,7 @@ function CartPage() {
     descriptionMessage,
     isDisabled,
+    selectedProducts,
   } = useCart();
```
어차피 cart가 context를 의존하고 있는데,
이걸 다시 props로 내려줄 필요가 있을까 싶어요..!
그냥 관련 컴포넌트가 context와 관련된 로직을 직접 호출하도록 하는게 코드의 흐름을 파악할 때 더 좋을 것 같네요..!

ShinjungOh replied:
장바구니와 주문확인 페이지 간 데이터 흐름을 context를 이용해 개선했어요.
전체 상품 대신에 선택된 아이템 ID만 전달하도록 했어요.
주문확인 페이지에서 ID를 기반으로 context에서 제품 정보 조회하는 방식으로 props drilling을 제거했어요.

[refactor: CartPage에서 OrderConfirmPage로 ID만 전달하도록 변경](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/a7fd0daa8fac254d6434c42e875660934fd79ea0)

JunilHwang commented on src/pages/OrderConfirmPage.tsx:
```diff
@@ -1,25 +1,104 @@
+    toggleCouponSelection,
+    isCouponSelected,
+    applyCoupons,
+  } = useCouponSelection();
```
앞선 리뷰와 동일합니다.

JunilHwang commented on src/utils/couponAvailability.ts:
```diff
@@ -0,0 +1,33 @@
+    const endMinute = Number(endTimeParts[1]);
+
+    const startTime = startHour * 60 + startMinute;
+    const endTime = endHour * 60 + endMinute;
```
이 코드도 묶어서 함수로 표현해주면 어떨까요?

ShinjungOh replied:
해당 로직을 별개의 함수로 분리했어요.
코드 양도 줄고 더 깔끔해졌네요!👍

[refactor: isCouponAvailable 유틸 함수의 시간 변환 로직을 함수로 분리](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/7edafe8d18b5a71e81c98416ce84ca820c30d3f2)

JunilHwang commented on src/utils/couponCalculations.ts:
```diff
@@ -0,0 +1,103 @@
+  }, 0);
+};
+
+const calculateSingleCouponDiscount = (
+  coupon: Coupon,
+  products: CartProduct[],
+  total: number,
+  shippingFee: number,
+): number => {
+  switch (coupon.discountType) {
+    case 'fixed':
+      return calculateFixed(coupon);
+
+    case 'percentage':
+      return calculatePercentage(coupon, total);
+
+    case 'buyXgetY':
+      return calculateBOGO(coupon, products);
+
+    case 'freeShipping':
+      return calculateFreeShipping(coupon, total, shippingFee);
+
+    default:
+      return 0;
+  }
+};
```
type이 100개 정도 된다고 가정해보면 이 코드를 유지보수 하기가 어렵지 않을까요!?

가령, DB에 할인쿠폰과 할인 로직에 대한 정보가 들어있다고 가정해보고 코드를 작성해보시면 어떨까 싶어요!

JunilHwang commented on src/utils/dateTimeFormatter.ts:
```diff
@@ -0,0 +1,25 @@
+  const endFormatted = `${endAMPM} ${endHour % 12 || 12}시`;
+
+  return `${startFormatted} ${endFormatted}까지`;
+};
```
12를 상수로 만들어서 관리하면 어떨까요?

ShinjungOh replied:
그러고보니 12가 자주 사용되는군요... 12를 상수로 분리했어요👍

[refactor: 오전 오후 기준시 12를 상수화](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/18252e0f6af2a08f0a98b56d66789b1f5a33b5c7)

JunilHwang commented on test/CartPage.test.tsx:
```diff
@@ -15,180 +12,178 @@ vi.mock('react-router', async () => {
   },
 }));
 
 describe('CartPage', () => {
```
전체적으로 훅에 대한 테스트가 아예 작성되지 않았어요.
지금처럼 복잡한 로직이 많을 때는 테스트가 더 촘촘하고 꼼꼼하게 작성되어 있어야 한답니다..!

JunilHwang reviewed (changes requested):
안녕하세요 루나~
장바구니 2단계 잘 진행해주셨네요! 고생하셨습니다!

---

전체적으로 살펴봤는데, 제일 개선이 필요한 부분은 context 사용에 대한 정책 같아요!
가령, useCart 가 context를 의존하고 있고, 다시 이 로직들이 props를 통해 하위 컴포넌트들에게 계속 전파되고 있습니다.
어차피 context를 사용해야 한다면, 이런 모습으로 사용하는건 오히려 복잡도만 증가시킨다고 생각해요! context의 이점을 살리지 못한다고 해야하나... 제가 느끼기에는 그렇습니다.

그리고 useData를 통해 한 종류의 데이터만 관리하고 있는데요, 만약 여러 종류의 데이터를 한 페이지에서 관리해야 한다면 어떻게 해야 좋을까요!?

---

> 훅의 반환값은 UI에 필요한 최소한의 정보만 포함하고, 내부 구현은 외부에 드러나지 않도록 캡슐화한다는 요구사항을 충족하고 있는지 궁금해요.

캡슐화는 잘 되고 있따고 생각합니다! 다만 OrderConfirmPage 의 경우 로직이 컴포넌트 내부에 있는데요, 이런 것들도 훅으로 분리해주면 어떨까 싶어요!

> OrderConfirmPage 컴포넌트에 여러 훅에서 불러온 값들이 있는데, 여기에서 사용되는 로직도 useOrderConfirm 훅 등으로 분리해주는 것이 좋을까요? 테스트 코드를 작성하려면 통합된 훅을 만드는 것이 좋아보이는데, 훅이 너무 많아지는 것 같아서 궁금해졌어요!

위에서 답변을 했는데, 바로 다음 질문이 이와 관련된 부분이군요 ㅎㅎ
훅은 지금보다 2~3배 정도 더 많아야 한다고 생각합니다.
많아서 나쁠게 있을까요? 컴포넌트 내부의 로직을 훅으로 분리해서 더 명확하게 관리하는거라서 오히려 좋다고 생각합니다.

컴포넌트에 있어야 하는 상태는 UI와 연관된 일부 상태 정도라고 생각해요.
나머지 어플리케이션과 직결되는 것들은 다 훅고 함수로 분리해서 관리되어야 사이드 이펙트를 감지하고 관리할 수 있지 않을까요!?

---

고지가 코앞입니다! 조금만 더 힘내봅시다!!

ShinjungOh committed:
refactor: apiClient 함수 추상화

ShinjungOh committed:
refactor: 불필요 리턴 타입 제거

ShinjungOh committed:
refactor: 배송비 관련 상수를 코드에 적용

ShinjungOh committed:
refactor: 오전 오후 기준시 12를 상수화

ShinjungOh committed:
refactor: 쿠폰 모달 여닫는 불필요 상태 로직을 값으로 개선

ShinjungOh committed:
refactor: isCouponAvailable 유틸 함수의 시간 변환 로직을 함수로 분리

ShinjungOh committed:
refactor: OrderConfirmPage의 로직을 useOrderConfirm 훅으로 분리

ShinjungOh committed:
refactor: 쿠폰 할인 금액의 음수 처리 로직 단순화

ShinjungOh committed:
refactor: CartOrderItem의 props명을 product에서 cart로 변경

ShinjungOh committed:
refactor: 관련성이 없는 useCartUI 내부의 로직을 useCart로 이동

ShinjungOh committed:
refactor: CartInfo에 Emotion css prop 적용으로 스타일링 개선

ShinjungOh committed:
refactor: CartPrice의 price prop 타입을 유연하게 변경

ShinjungOh committed:
refactor: 쿠폰 모달 상태 관리 구조 개선

ShinjungOh committed:
test: CartPage 테스트에서 descriptionMessage를 문자열로 변경

ShinjungOh committed:
refactor: Context 활용해 props drilling 해결 및 데이터 흐름 개선

- useCartData 삭제 및 useCartItemsData로 대체
- CartList, CartMain 컴포넌트의 props drilling 제거
- useCartSelection, useCartCalculations의 매개변수 의존성 제거

ShinjungOh committed:
refactor: CartPage에서 OrderConfirmPage로 ID만 전달하도록 변경

ShinjungOh committed:
fix: 장바구니 페이지 초기 로딩 시 데이터가 표시되지 않는 문제 해결

ShinjungOh commented:
# 🛒 장바구니 - 리팩토링

안녕하세요 해먼드! 피드백 반영을 마치고 리뷰 요청 드립니다!!

마감 시간이 별로 없어서 우선순위가 높은 사항 위주로 리팩토링을 진행했어요.
모달과 관련한 로직(useFocusTrap 등)은 장바구니 미션의 학습 목표와는 조금 거리가 있어, 나머지 주요 리뷰를 우선적으로 반영한 점 양해 부탁드립니다..!

어느덧 마지막 미션의 마지막 리뷰라니 감회가 새롭고 싱숭생숭 하네요..
그동안 해먼드와의 리뷰 덕분에 많이 배웠습니다. 감사합니다!
남은 리뷰도 확인하고, 특히 테스트 관련해서도 꼭 살펴보도록 할게요!!💪🔥

리팩토링하며 변경된 사항을 정리해서 남겨볼게요!

[🔗 배포 사이트](https://shinjungoh.github.io/react-shopping-cart/)

<br>

## 기능 영상 



https://github.com/user-attachments/assets/d47007eb-c876-4557-b20d-67149bbb2fd4



<br>

## 테스트 통과 

![CleanShot 2025-06-13 at 10 13 10@2x](https://github.com/user-attachments/assets/9c277887-e8e3-4697-8d2a-1d4c86470607)

<br>

## ⚙️ 주요 변경 사항

### 1. 상태 관리 및 데이터 처리

context를 사용해 Props drilling을 제거했어요. 
OrderConfirmPage 컴포넌트에 여러 훅에서 불러온 값들이 있는데, 여기에서 사용되는 로직도 useOrderConfirm 훅으로 분리했어요.

[refactor: OrderConfirmPage의 로직을 useOrderConfirm 훅으로 분리](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/6143396688280fe9ac9c7413cc24b12c8519f732)
[refactor: 관련성이 없는 useCartUI 내부의 로직을 useCart로 이동](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/296294f4d720654419919a460e7574cb552f8d45)
[refactor: Context 활용해 props drilling 해결 및 데이터 흐름 개선](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/37a2212822ff662ba0939e3500ef84066c16149f)
[refactor: CartPage에서 OrderConfirmPage로 ID만 전달하도록 변경](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/a7fd0daa8fac254d6434c42e875660934fd79ea0)
[fix: 장바구니 페이지 초기 로딩 시 데이터가 표시되지 않는 문제 해결](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/6e072447bdc4c6012ee99d645e7d2db24a62599f)


### 2. 쿠폰&할인 관련 개선

[refactor: 쿠폰 모달 여닫는 불필요 상태 로직을 값으로 개선](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/5c61b99f325585fb10307517376919705ddb3648)
[refactor: 쿠폰 모달 상태 관리 구조 개선](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/f654e76472555af34d82d0af6c6ed9e4ba6babde)
[refactor: 쿠폰 할인 금액의 음수 처리 로직 단순화](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/2326f80ee4243f4a999278164fb9284ba5ba45dc)


### 3. UIUX 개선


[refactor: CartInfo에 Emotion css prop 적용으로 스타일링 개선](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/853dc09c1375cd4c69e020e664334db2b41e6d0c)
[refactor: CartPrice의 price prop 타입을 유연하게 변경](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/91af7d51ca232faef3c02c52730ba2fff22dcb59)
[refactor: CartOrderItem의 props명을 product에서 cart로 변경](https://github.com/woowacourse/react-shopping-cart/pull/405/commits/7069f975b573953343fc7662039a09c3d4308160)


### 4. 기타 구현사항

그 밖의 자잘한 리팩토링을 진행했어요.

- API&데이터 관련 코드 개선
- 매직 넘버 상수화
- isCouponAvailable 유틸 함수의 시간 변환 로직을 함수로 분리 
- 테스트 통과되도록 코드 개선

JunilHwang commented on src/hooks/useCart.ts:
<img width="658" alt="image" src="https://github.com/user-attachments/assets/89b1dc1f-8718-499e-a9fe-5fd28a34e066" />

지금 훅에 대한 테스트가.. 하나밖에 없네요 ㅠㅠ 

컴포넌트는 테스트하기도 어렵고, 정확도가 생각보다 높지 않습니다.
데이터가 있고, 데이터의 변경을 발생시키는 함수를 모아서 관리하는게 훅이라고 생각해요.
순수 함수는 "데이터의 변경"위주라고 한다면, 훅은 "상태"와 "함수"의 조합인거죠.
그래서 훅에 대한 테스트를 작성하면 어느정도 어플리케이션의 다양한 규칙에 대해 검증이 가능하다고 생각합니다.

JunilHwang commented on src/components/Cart/CartMain.tsx:
```diff
@@ -1,57 +1,25 @@
-      <CartList checkedItems={checkedItems} setCheckedItems={setCheckedItems} />
-
-      <CartInfo>
-        <InfoIconImage src={infoIcon} alt="infoIcon" />
-        <p>
-          총 주문 금액이 {SHIPPING_FEE_THRESHOLD.toLocaleString()}원 이상일 경우 무료 배송됩니다.
-        </p>
-      </CartInfo>
-
+      <CartList />
+      <CartInfo
+        description={`총 주문 금액이 ${SHIPPING_FEE_THRESHOLD.toLocaleString()}원 이상일 경우 무료 배송됩니다.`}
+      />
       <CartFooter price={price} shippingFee={shippingFee} totalPrice={totalPrice} />
     </>
   );
```
```suggestion
  const selection = useCartSelection();
  const cartCalculations = useCartCalculations({ checkedIds: selection.checkedItems });

  return (
    <>
      <SelectAllBox checked={selection.isAllChecked} onChange={selection.checkAll} />
      <CartList />
      <CartInfo
        description={`총 주문 금액이 ${SHIPPING_FEE_THRESHOLD.toLocaleString()}원 이상일 경우 무료 배송됩니다.`}
      />
      <CartFooter {...cartCalculations} />
    </>
  );
```
이렇게 표현할 수 있을 것 같아요!

JunilHwang commented on src/utils/couponCalculations.ts:
```diff
@@ -0,0 +1,103 @@
+  }, 0);
+};
+
+const calculateSingleCouponDiscount = (
+  coupon: Coupon,
+  products: CartProduct[],
+  total: number,
+  shippingFee: number,
+): number => {
+  switch (coupon.discountType) {
+    case 'fixed':
+      return calculateFixed(coupon);
+
+    case 'percentage':
+      return calculatePercentage(coupon, total);
+
+    case 'buyXgetY':
+      return calculateBOGO(coupon, products);
+
+    case 'freeShipping':
+      return calculateFreeShipping(coupon, total, shippingFee);
+
+    default:
+      return 0;
+  }
+};
+
```
지금은 명령형으로 코드가 작성되어있는데, 이걸 선언형으로 표현하는 방법에 대해서 고민해보시면 좋겠어요..!

JunilHwang commented on src/pages/OrderConfirmPage.tsx:
```diff
@@ -1,25 +1,103 @@
+    price: orderPrice,
+  } = useOrderConfirm();
+
+  const handleModalOpen = () => {
+    setIsModalOpen(true);
+  };
+
+  const handleModalClose = () => {
+    setIsModalOpen(false);
+  };
+
+  const handleConfirmCoupon = (coupons: Coupon[]) => {
+    setSelectedCoupons(coupons);
+    setIsModalOpen(false);
+  };
```
```suggestion
  const openModal = () => {
    setIsModalOpen(true);
  };

  const closeModal = () => {
    setIsModalOpen(false);
  };

  const applyCopuons = (coupons: Coupon[]) => {
    setSelectedCoupons(coupons);
    setIsModalOpen(false);
  };
```

함수가 하는 일로 이름을 지어주세요!

JunilHwang commented on src/components/Modal/CouponModal.tsx:
```diff
@@ -0,0 +1,218 @@
+}
+
+function CouponModal({
+  isOpen,
+  onClose,
+  onConfirm,
+  initialCoupons = [],
+  products,
+  subtotal,
+  shippingFee,
+}: CouponModalProps) {
+  const [selectedCoupons, setSelectedCoupons] = useState<Coupon[]>(initialCoupons);
+
+  const tempCouponDiscount = calculateCouponDiscount({
+    coupons: selectedCoupons,
+    products,
+    total: subtotal,
+    shippingFee,
+  });
+
+  const toggleCouponSelection = (coupon: Coupon) => {
+    if (!isCouponAvailable(coupon)) {
```
여기에서 보면
isOpen => 값으로 쓰임
isCouponAvailable => 함수로 쓰임

이렇습니다.

isOpen(modal) // modal이 open 상태인지 확인하는 함수가 있다면?

isOpen이라는 이름은 이렇게도 쓰일 수 있을 것 같아요.

그래서 이럴 땐 isOpen 보단 opened 처럼 표현하면 어떨까 싶어요!

JunilHwang commented on src/components/Modal/CouponModal.tsx:
```diff
@@ -0,0 +1,218 @@
+    }
+
+    setSelectedCoupons((prev) => {
+      const isSelected = prev.some((item) => item.id === coupon.id);
+
+      if (isSelected) {
+        return prev.filter((item) => item.id !== coupon.id);
+      }
```
앞선 리뷰와 동일합니다.
```tsx
const isSelected = (target) => target.some(item => item.id === coupon.id)
```
이렇게 똑같은 이름의 함수가 있을 수 있어요.

JunilHwang commented on src/components/Modal/CouponModal.tsx:
```diff
@@ -0,0 +1,218 @@
+    });
+  };
+
+  const isCouponSelected = (couponId: number) => {
+    return selectedCoupons.some((coupon) => coupon.id === couponId);
+  };
```
ㅎㅎ 바로 유사한 사례가 여기 있네요

JunilHwang commented on src/components/Modal/CouponModal.tsx:
```diff
@@ -0,0 +1,218 @@
+        />
+        <CouponListContainer>
+          {COUPONS.map((coupon) => {
+            const isAvailable = isCouponAvailable(coupon);
+            const isSelected = isCouponSelected(coupon.id);
```
흠... is 라는 prefix를 값에 사용하는게... 맞을까요?
이런 코드가 앞으로도 계속 나올 것 같네요.

JunilHwang reviewed (approved):
안녕하세요 루나!
리팩토링 잘 진행해주셨네요 ㅎㅎ 고생하셨습니다.

반복적으로 남기는 이야기인데,
boolean에 대한 네이밍 때문에 코드를 작성하는 사람도 보는 사람도 헷갈린다고 생각해요.

isSelected
isCouponSelected
isAvailable
isCouponAvailable
isSelected
isCouponSelected
isSelectable

이게 다 루나가 만들어주신 것들인데요. 어떤게 함수인지 변수인지 구분하실 수 있을까요..?
- const isSelected = isCouponSelected(...)
- const isCouponSelected = isSelected(...)

이름에는 일관성이 있어야 한다고 생각해요. 그래야 코드를 읽을 때도 작성할 때도 고민을 덜 하게 되니까요. boolean 값에 is prefix를 붙이는건 이런 일관성을 해치는 네이밍이라고 생각합니다.

이 문제를 어떻게 해결할 수 있을지 고민하시면 좋겠어요.
루나가 작성해주신 코드만 봐도 스스로 "이걸 어떻게 네이밍해야 좋을까? 이름이 다 겹치네..?" 라는 고민하다가 지금의 모습이 된 것 같네요 ㅎㅎ

그리고 지금 컴포넌트를 제외하고 테스트 해보면 커버리지가 굉장히 낮아요. 사실 훅에 대한 테스트는 없다고 봐도 무방한 상태입니다... ㅠㅠ

프론트엔드 개발을 할 때 구성 요소는 다음과 같다고 생각해요.
1. UI
2. Data
3. Function

컴포넌트는 UI, data, function 의 조합이고
훅은 data, function 의 조합으로 사용 가능합니다.

ui는 data를 통해 표현돼죠.

다르게 생각하자면 data의 표현 수단이 ui라고도 생각할 수 있어요.

그렇다면 우리는 ui에 대해 테스트하지 않아도, 함수와 함수의 결과인 data에 대해 검증할 수 있다면 ui에 대한 검증은 충분히 커버할 수 있다고 생각해요.

특히 지금 처럼 복잡한 로직을 다루는 어플리케이션의 경우에는 더더욱!
그래서 웬만하면 훅으로 분리해서 테스트 하시는걸 추천드립니다 ㅎㅎ

마지막으로 useData를 사용하는 구간이 있는데요, 이걸 fetch 를 하는 시점에 캐시해서 관리하면 context를 사용할 필요가 아예 없어요. 간단한 예시를 보여드리자면,

```tsx
const  cache: Record<string, unknown> = {};
export function useData<T>({ fetcher, name }: { fetcher: () => Promise<T>; name: string }) {
  const [data, setData] = useState(() => cache[name] as T);

  const request = useCallback(() => {
    fetcher().then((res) => {
      cache[name] = res;
      setData(res);
    });
  }, [fetcher, name, setData]);

  useEffect(() => {
    const hasData = cache[name];

    if (hasData) {
      return;
    }
    request();
  }, [data, name, request]);

  return { data, refetch: request };
}
```

이렇게 가능합니다 ㅎㅎ 
이를 통해 드리고 싶은 이야기는, 앞으로 리액트를 계속 사용하게 되겠지만 리액트라는 틀에 갇혀서 생각할 필요가 없다는 것입니다.
리액트가 아닌 바닐라로 만든다고 생각해보면 자연스럽게 객체 기반으로 캐시를 하는 방법을 생각하셨을 것 같아요.
리액트 이전에 학습했던 것들을 잘 인지해주세요!

---

마지막 리액트 미션의 마지막 리뷰는 여기서 마무리하도록 하겠습니다!
부족한 리뷰어였으나 잘 따라와주셔서 감사합니다 루나!
고생하셨습니다!
